File: java/src/chapter2/section1/E26_PrimitiveTypes.java
package chapter2.section1;

import edu.princeton.cs.algs4.StdOut;
import edu.princeton.cs.algs4.StdRandom;
import edu.princeton.cs.algs4.Stopwatch;

// For 100000 random numbers
//    Insertion is 7.2 times slower than InsertionPrimitive

public class E26_PrimitiveTypes {

  public static class InsertionPrimitive {
    public static void sort(int[] a) {
      int N = a.length;
      for (int i = 1; i < N; i++) {
        for (int j = i; j > 0 && less(a[j], a[j - 1]); j--) {
          exch(a, j, j - 1);
        }
      }
    }

    private static boolean less(int v, int w) {
      return v < w;
    }

    private static void exch(int[] a, int i, int j) {
      int t = a[i];
      a[i] = a[j];
      a[j] = t;
    }
  }

  public static class Insertion {
    public static <T extends Comparable<T>> void sort(T[] a) {
      int N = a.length;
      for (int i = 1; i < N; i++) {
        for (int j = i; j > 0 && less(a[j], a[j - 1]); j--) {
          exch(a, j, j - 1);
        }
      }
    }

    private static <T extends Comparable<T>> boolean less(T v, T w) {
      return v.compareTo(w) < 0;
    }

    private static <T extends Comparable<T>> void exch(T[] a, int i, int j) {
      T t = a[i];
      a[i] = a[j];
      a[j] = t;
    }
  }

  public static double time(String alg, Integer[] a) throws IllegalArgumentException {
    Stopwatch timer;
    switch (alg) {
      case "Insertion":
        timer = new Stopwatch();
        Insertion.sort(a);
        return timer.elapsedTime();
      case "InsertionPrimitive":
        int N = a.length;
        int[] b = new int[N];
        for (int i = 0; i < N; i++) {
          b[i] = a[i];
        }
        timer = new Stopwatch();
        InsertionPrimitive.sort(b);
        return timer.elapsedTime();
      default:
        throw new IllegalArgumentException("Unknown algorithm: " + alg);
    }
  }

  public static double timeRandomInput(String alg, int N, int T) {
    double total = 0.0;
    Integer[] a = new Integer[N];
    for (int t = 0; t < T; t++) {
      for (int i = 0; i < N; i++) {
        a[i] = StdRandom.uniformInt(N);
      }
      try {
        total += time(alg, a);
      } catch (IllegalArgumentException e) {
        StdOut.println(e.getMessage());
        return 0;
      }
    }
    return total;
  }

  public static void main(String[] args) {
    if (args.length != 4) {
      StdOut.println("Usage: java SortCompare <alg1> <alg2> <array size> <number of trials>");
      return;
    }

    String alg1 = args[0];
    String alg2 = args[1];
    int N = Integer.parseInt(args[2]);
    int T = Integer.parseInt(args[3]);
    double t1 = timeRandomInput(alg1, N, T);
    double t2 = timeRandomInput(alg2, N, T);

    if (t1 < 1e-6 || t2 < 1e-6) {
      StdOut.printf("The elapsed time for one or both algorithms is too small to make a reliable comparison.\n");
    } else {
      double ratio = t2 / t1;
      if (ratio > 1) {
        StdOut.printf("For %d random numbers\n   %s is %.1f times faster than %s\n", N, alg1, ratio, alg2);
      } else {
        StdOut.printf("For %d random numbers\n   %s is %.1f times slower than %s\n", N, alg1, 1 / ratio, alg2);
      }
    }
  }
}

File: typescript/chapter2/section1/E26_PrimitiveTypes.ts
import { performance } from "perf_hooks"
import { updateProgressBar } from "../../utils"

const lessPrimitive = (a: number, b: number): boolean => a < b

const lessNumberObject = (a: Number, b: Number): boolean => a.valueOf() < b.valueOf()

const swap = <T>(arr: T[], i: number, j: number): void => {
  const temp = arr[i]
  arr[i] = arr[j]
  arr[j] = temp
}

const sortPrimitive = (arr: number[]): void => {
  const length = arr.length
  for (let i = 1; i < length; i++) {
    for (let j = i; j > 0 && lessPrimitive(arr[j], arr[j - 1]); j--) {
      swap(arr, j, j - 1)
    }
  }
}

const sortNumberObject = (arr: Number[]): void => {
  const length = arr.length
  for (let i = 1; i < length; i++) {
    for (let j = i; j > 0 && lessNumberObject(arr[j], arr[j - 1]); j--) {
      swap(arr, j, j - 1)
    }
  }
}

const measureTimePrimitive = (array: number[]): number => {
  const startTime = performance.now()
  sortPrimitive(array)
  const endTime = performance.now()
  return (endTime - startTime) / 1000
}

const measureTimeNumberObject = (array: Number[]): number => {
  const startTime = performance.now()
  sortNumberObject(array)
  const endTime = performance.now()
  return (endTime - startTime) / 1000
}

const measureAverageTime = (
  size: number,
  trials: number
): { primitive: number; numberObject: number } => {
  const totalTimePrimitive = Array.from({ length: trials }).reduce((acc: number, _, t) => {
    const primitiveArray: number[] = Array.from({ length: size }, () =>
      Math.floor(Math.random() * size)
    )
    const time = measureTimePrimitive(primitiveArray)
    updateProgressBar(t + 1, trials, `Trial ${t + 1}/${trials}`)
    return acc + time
  }, 0) as number

  const totalTimeNumberObject = Array.from({ length: trials }).reduce((acc: number, _, t) => {
    const numberObjectArray: Number[] = Array.from(
      { length: size },
      () => new Number(Math.floor(Math.random() * size))
    )
    const time = measureTimeNumberObject(numberObjectArray)
    updateProgressBar(t + 1, trials, `Trial ${t + 1}/${trials}`)
    return acc + time
  }, 0) as number

  return {
    primitive: totalTimePrimitive / trials,
    numberObject: totalTimeNumberObject / trials,
  }
}

const main = (): void => {
  const args = process.argv.slice(2)

  if (args.length !== 2) {
    console.log("Usage: ts-node SortCompare <array size> <number of trials>")
    return
  }

  const size = parseInt(args[0])
  const trials = parseInt(args[1])

  const { primitive, numberObject } = measureAverageTime(size, trials)

  if (primitive < 1e-6 || numberObject < 1e-6) {
    console.log(
      "The elapsed time for one or both algorithms is too small to make a reliable comparison."
    )
  } else {
    const ratio = numberObject / primitive
    if (ratio > 1) {
      console.log(
        `For ${size} random numbers\n   sortPrimitive is ${ratio.toFixed(
          1
        )} times faster than sortNumberObject`
      )
    } else {
      console.log(
        `For ${size} random numbers\n   sortPrimitive is ${(1 / ratio).toFixed(
          1
        )} times slower than sortNumberObject`
      )
    }
  }
}

main()
